packages/react-instantsearch-hooks/src/hooks/__tests__/useInstantSearch.test.tsx

```diff
@@ -0,0 +1,458 @@
+      }
+
+      const { container } = render(
+        <InstantSearchHooksTestWrapper>
+          <App />
+        </InstantSearchHooksTestWrapper>
+      );
+
+      await wait(0);
+
+      expect(container.innerHTML).toMatchInlineSnapshot(
+        `"<button type=\\"button\\"></button>"`
+      );
+
+      userEvent.click(container.querySelector('button')!);
+
+      await wait(0);
+
+      expect(container.innerHTML).toMatchInlineSnapshot(
+        `"<button type=\\"button\\">hey</button>hey"`
+      );
+    });
+  });
+
+  describe('middleware', () => {
+    test('gives access use function', async () => {
+      const wrapper = createInstantSearchTestWrapper();
+
+      const subscribe = jest.fn();
+      const onStateChange = jest.fn();
+      const unsubscribe = jest.fn();
+      const middleware = jest.fn(() => ({
+        subscribe,
+        onStateChange,
+        unsubscribe,
+      }));
+
+      const { result, waitForNextUpdate, unmount } = renderHook(
+        () => {
+          const { use, search } = useInstantSearch();
+          useEffect(() => use(middleware), [use]);
+
+          return search;
+        },
+        {
+          wrapper: ({ children }) =>
+            wrapper({
+              children: (
+                <>
+                  <SearchBox />
+                  {children}
+                </>
+              ),
+            }),
+        }
+      );
+
+      // augmented once middleware was used
+      expect(result.current.middleware).toEqual([
+        {
+          creator: middleware,
+          instance: {
+            onStateChange,
+            subscribe,
+            unsubscribe,
+          },
+        },
+      ]);
+
+      expect(subscribe).toHaveBeenCalledTimes(1);
+      expect(onStateChange).toHaveBeenCalledTimes(0);
+      expect(unsubscribe).toHaveBeenCalledTimes(0);
+      expect(middleware).toHaveBeenCalledTimes(1);
+
+      await waitForNextUpdate();
+
+      expect(subscribe).toHaveBeenCalledTimes(1);
+      expect(onStateChange).toHaveBeenCalledTimes(0);
+      expect(unsubscribe).toHaveBeenCalledTimes(0);
+
+      // simulate a change in query
+      result.current.renderState.indexName.searchBox!.refine('new query');
+      await waitForNextUpdate();
+
+      expect(subscribe).toHaveBeenCalledTimes(1);
+      expect(onStateChange).toHaveBeenCalledTimes(1);
+      expect(unsubscribe).toHaveBeenCalledTimes(0);
+
+      unmount();
+
+      expect(subscribe).toHaveBeenCalledTimes(1);
+      expect(onStateChange).toHaveBeenCalledTimes(1);
+      // unsubscribe is first called by the parent InstantSearch unmounting
+      // 🚨 dispose doesn't remove middleware (because otherwise routing would break)
+      // then unuse is called by this widget itself unmounting.
+      // if only the component with useInstantSearch is unmounted, unsubscribe is called once.
+      expect(unsubscribe).toHaveBeenCalledTimes(2);
+
+      // middleware was removed
+      expect(result.current.middleware).toEqual([]);
```

Comment:
This is testing InstantSearch.js, not the public API that it exposes. So I don't think we should have it here.